[week5] 고양이 장난감가게 유효성 추가하기, 회원 관리 만들기

caoyu-dev opened the pull request:
과제 목표
--

1. 고양이 장난감가게 유효성 검사 추가하기

- [x] 고양이 장난감 목록 얻기 - `GET /products`
- [x] 고양이 장난감 상세 조회하기 - `GET /products/{id}`
- [x] 고양이 장난감 등록하기 - `POST /products`
- [x] 고양이 장난감 수정하기 - `PATCH /products/{id}`
- [x] 고양이 장난감 삭제하기 - `DELETE /products/{id}`

2. 회원 관리하기

- [x] 회원 생성하기 - `POST /user`
- [x] 회원 수정하기 - `POST /user/{id}`
- [x] 회원 삭제하기 - `DELETE /user/{id}`


요구 사항
--
- [x] 테스트 커버리지 100%를 달성해야 합니다.
- [x] 모든 API 테스트를 통과해야 합니다.
- [x] 모든 E2E 테스트를 통과해야 합니다.

제한 조건
--
단, 모든 기능을 TDD 로 구현할 것

----
늦어져서 죄송합니다. 리뷰 부탁드립니다. 
이번 주는 먼저 완성하고 보완하는 방향으로 진행하겠습니다.

caoyu-dev committed:
feat: [POST] /users 요청에 대한 UserControllerWebTest 작성
해당 webTest 를 바탕으로 User, UserData, UserController, UserService, UserRepository 순서대로 설정

caoyu-dev committed:
refactor: EOF 에 newline 추가

caoyu-dev committed:
feat: build.gradle 에 dozermapper 추가

caoyu-dev committed:
feat: User 클래스 구현

caoyu-dev committed:
feat: UserCreateDate 클래스 구현
UserDto 세분화

caoyu-dev committed:
feat: UserController::createUser 구현

caoyu-dev committed:
feat: UserService::createUser 구현
//TODO: dozermapper 활용은 모든 설정 이후에 진행

caoyu-dev committed:
feat: UserRepository 인터페이스 구현

caoyu-dev committed:
feat: JpaUserRepository 인터페이스 생성

caoyu-dev committed:
refactor: [POST] /users 요청에 대한 UserControllerWebTest 수정
1. 메서드명 create -> createWithValidAttributes
2. 메서드 createWithInvalidAttributes 추가

caoyu-dev committed:
feat: UserNotFoundException 생성

caoyu-dev committed:
refactor: UserController::createUser @Valid 추가

caoyu-dev commented:
`dozermapper` 의 경우 모든 테스트 작성이 끝나고 리펙터링 과정에서 진행하도록 하겠습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/UserNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class UserNotFoundException extends RuntimeException {
+    public UserNotFoundException(Long id) {
+        super("Product not found: " + id);
```
`Product`라고 되어있군요. User로 변경해야 될 것 같아요

caoyu-dev replied:
아핫 복붙했더니 이런 폐해가… 잡아주셔서 감사합니다

hannut91 reviewed (commented):
잘하셨습니다 ㅎㅎ 의식적으로 테스트를 먼저 작성해 보는 연습을 해봅시다

caoyu-dev committed:
refactor: 출력문 수정

caoyu-dev committed:
refactor: @NotBlank 추가
UserControllerWebTest::createWithInvalidAttributes 실패 이유가 @NotBlank 추가가 안되어서 그런 거였음

caoyu-dev committed:
feat: [PUT/PATCH] /users/1 요청에 대한 UserControllerWebTest 작성

caoyu-dev committed:
feat: UserController::update 구현 및 메서드명 변경
createUser -> create

caoyu-dev committed:
feat: UserService::updateUser 구현

caoyu-dev committed:
feat: User 클래스 수정
update 메서드 추가

caoyu-dev committed:
feat: UserUpdateData 클래스 구현
//TODO
UserCreateData 와 동일한 구조라서 어떻게 변경할건지 고민

caoyu-dev committed:
feat: UserRepository 인터페이스 수정
findById 추가

caoyu-dev committed:
feat: JpaUserRepository 인터페이스 수정
findById 추가

caoyu-dev committed:
refactor: @ResponseStatus(HttpStatus.NOT_FOUND) 추가
UserControllerWebTest::updateWithNotExistedProduct 실패 원인 정확하지 않음
여러 시도를 한 결과 커밋한 내용으로 진행하니 성공!

caoyu-dev committed:
refactor: 메서드명 수정
updateWithNotExistedProduct -> updateWithNotExistedUser

caoyu-dev committed:
feat: [DELETE] /users/1 요청에 대한 UserControllerWebTest 작성

caoyu-dev committed:
feat: UserController::delete 구현

caoyu-dev committed:
feat: UserSerivce::deleteUser 구현

caoyu-dev committed:
feat: UserRepository 인터페이스 수정
delete 추가

caoyu-dev committed:
feat: JpaUserRepository 인터페이스 수정
delete 추가

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,55 @@
+@Service
+public class UserService {
+    private final UserRepository userRepository;
+    private User user;
```
이 `user`는 안쓰이는 것 같네요

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,55 @@
+    private User getUser(Long id) {
+        User user = userRepository.findById(id)
+                .orElseThrow(() -> new UserNotFoundException(id));
+        return user;
```
메서드 명과 메서드의 반환타입으로 응답값이 무엇인지 추측할 수 있고, 그리고 `findById`로 인해서 어떤 값을 반환하는지 충분히 추론이 가능해서 좌변에 변수가 없어도 이해하는데 어렵지 않을 것 같아요. 따라서 바로 반환해줘도 될 것 같습니다

caoyu-dev committed:
refactor: APPLICATION_JSON_UTF8 제거

caoyu-dev committed:
refactor: 불필요한 import 삭제

caoyu-dev committed:
refactor: @ResponseStatus(HttpStatus.NOT_FOUND) 추가

caoyu-dev committed:
refactor: UserService::getUser inline

caoyu-dev committed:
refactor: delete 메서드 Long id -> User user

caoyu-dev committed:
refactor: delete 메서드 Long id -> User user

caoyu-dev committed:
refactor: lombok 버전 변경

caoyu-dev committed:
feat: @Utf8WebTest 생성
한글 인코딩 builder 타입 제외하고 애노테이션이 가능한 형태로 변경

caoyu-dev committed:
refactor: builder 로 들어간 기존의 한글 인코딩 삭제

caoyu-dev committed:
refactor: playwright 버전 변경
1.8.1 -> 1.21.0

caoyu-dev committed:
refactor: mode: 'development' 추가

caoyu-dev commented:
6/1(수) 
1. `UserControllerWebTest` 에서 회원 생성, 수정, 삭제 테스트
2. `ProductControllerWebTest` 에 대한 `./gradlew run` 오류, `npm install` 오류
이때 `./gradlew run` 오류는 java 버전 차이 때문에 발생한 오류이고, java 11 -> 17 -> 15 설치 결과 (해결)

* java 17 - lombok 1.18.16 호환 문제가 생겨 lombok 1.18.22 으로 업글
* `'mediatype.application_json_utf8' is deprecated` 해결하기 위해 builder 차용

caoyu-dev commented:
6/2(목)
1. 전날 발생한 `npm install` 오류는 디스코드에 `전계동` 님이 작성해주신 글을 바탕으로 해결했다. 

> e2e 테스트에서 npm install 실행시,
>
>npm ERR! Error: Failed to download chromium, caused by
>npm ERR! Error: ERROR: Playwright does not support chromium on mac12
>
>위와 같은 에러 날 경우, 
>
>package.json에서   "playwright": "^1.21.0",
>최신 버전으로 변경하니 잘 설치됨.
>
>제 경우에는 npm ci로 해도 안 되었고요.ㅜㅜ 버전 변경하니 되네요. 


2. 하기의 아샬 님 말씀에 따라 builder 를 버리고 `@Utf8WebTest` 을 만드는 형태로 전환 https://github.com/CodeSoom/spring-week5-assignment-1/pull/65/commits/e2846bf2f21dd6d5a3f7645bbcc2aeb5a5c050ed

>특별한 경우가 아니면 빌더를 따로 쓰는 걸 추천하지 않습니다. 장식 코드가 늘어나는 경향이 있어요. 예전에는 BeforeEach에서 mockMvc를 만들어서 썼지만, 지금은 그러지 말라고 어노테이션으로 지원해주는 거죠.
>
>이 문제는 영어권에서 CJKV 사용자의 편의성을 무시하고 결정한 게 원인이라고 보는데, 한쪽은 UTF8이 기본이라면서 >MediaType.APPLICATION_JSON_UTF8를 deprecated시키고, 한쪽은 인코딩 처리를 정리하면서 UTF8을 기본에서 빼면서 앞의 결정과 충돌이 나게 된 거죠.


3. `This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.` 오류 발생 (해결중)


4. `e2e` 테스트 오류 (해결중)

>e2e 테스트는 
>설치
>$ cd web
>$ npm install
>
>
>웹 실행
>$ npm run start
>
>
>웹 테스트 - 다른 터미널창에서 실행
>$ npm run e2e
>
>
>해야 정상적으로 잘 되네요.

현재 network error 가 발생한다. (피곤)

caoyu-dev committed:
refactor: 파일명 수정
UserControllerWebTest 와 통일성을 위해 바꿈

caoyu-dev committed:
feat: UserServiceTest 생성

caoyu-dev committed:
refactor: @Transactional 추가

caoyu-dev committed:
chore: update 정리

caoyu-dev committed:
refactor: ErrorResponse handler 수정

caoyu-dev committed:
refactor: NotFoundException 수정
NotFoundErrorAdvice 클래스에서 @ResponseStatus(HttpStatus.NOT_FOUND) 를 받고 있으므로 제거

caoyu-dev committed:
chore: 삭제된거 복구

caoyu-dev committed:
feat: UserTest 생성

caoyu-dev committed:
refactor: UserControllerWebTest 에 @Utf8WebTest 추가

caoyu-dev committed:
refactor: @NotBlank 로 수정
API 테스트에서 userUpdateData 의 email 이 rejected value [null]

caoyu-dev committed:
refactor: UserController::update 메서드 @PatchMapping 으로 변경

caoyu-dev committed:
refactor: User 클래스 @NotBlank 삭제
FAILURE Status: 400
139
      JSON: {
140
          "timestamp": "2022-06-02T15:56:37.697+00:00",
141
          "status": 400,
142
          "error": "Bad Request",
143
          "trace": "org.springframework.web.bind.MethodArgumentNotValidException: Validation failed for argument [1] in public com.codesoom.assignment.domain.User com.codesoom.assignment.controllers.UserController.update(java.lang.Long,com.codesoom.assignment.dto.UserUpdateData): [Field error in object 'userUpdateData' on field 'email': rejected value [null]; codes [NotBlank.userUpdateData.email,NotBlank.email,NotBlank.java.lang.String,NotBlank]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [userUpdateData.email,email]; arguments []; default message [email]]; default message [must not be blank]] \n\tat

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,107 @@
+
+    @Test
+    void updateUserWithExistedId() {
+        UserUpdateData userUpdateData = UserUpdateData.builder()
+                .name("meow")
+                .email("choyumin01@gmail.com")
+                .password("meow1234")
+                .build();
```
이 데이터도 여기서 만들 필요 없이, 미리 만들어놔도 될 것 같아요. 왜냐하면 반복적으로 데이터를 만들고 있고, 이 데이터를 만드는 과정은 이 테스트에서 중요한 관심사가 아니기 때문이에요

hannut91 reviewed (commented):
API테스트 결과를 보니, 몇 개가 실패하는 것 같네요. API 테스트를 통해서 확인해 봅시다

caoyu-dev committed:
refactor: UserUpdateData 수정
email @NotBlank 삭제 및 name, password @NotEmpty 수정

caoyu-dev committed:
refactor: Controller 에 @CrossOrigin 추가 및 ProductData 에 @Mapping 추가
e2e 오류 해결

caoyu-dev commented:
6/4(토)
1. e2e network error 는 `@CrossOrigin` 를 추가함으로써 해결
> 6/2(목)
> 4\. `e2e` 테스트 오류 (해결중)

2. ~~JavaDoc 작성~~
3. ~~Test 를 given-when-then 패턴으로 재작성~~ (패스)
4. ~~__given() 공부__~~

caoyu-dev committed:
chore: JavaDoc 생성
